File: frontend/src/data/properties.ts
import { API_BASE } from "../api";
import type { PropertyCard } from "../types";

// ---------------------------------------------------------------------------
// Image fallback pools (used when Zillow photo is unavailable)
// ---------------------------------------------------------------------------
const oneBedModules = import.meta.glob(
  "../assets/HouseImages/1BedImages/*.{jpg,png,jpeg,webp}",
  { eager: true }
);
const twoBedModules = import.meta.glob(
  "../assets/HouseImages/2BedImages/*.{jpg,png,jpeg,webp}",
  { eager: true }
);
const threeBedModules = import.meta.glob(
  "../assets/HouseImages/3BedImages/*.{jpg,png,jpeg,webp}",
  { eager: true }
);
const fourBedModules = import.meta.glob(
  "../assets/HouseImages/4BedImages/*.{jpg,png,jpeg,webp}",
  { eager: true }
);

const ONE_BED_IMAGES = Object.values(oneBedModules).map((mod: any) => mod.default);
const TWO_BED_IMAGES = Object.values(twoBedModules).map((mod: any) => mod.default);
const THREE_BED_IMAGES = Object.values(threeBedModules).map((mod: any) => mod.default);
const FOUR_BED_IMAGES = Object.values(fourBedModules).map((mod: any) => mod.default);

function getFallbackImage(beds: number, index: number): string {
  switch (beds) {
    case 1: return ONE_BED_IMAGES[index % ONE_BED_IMAGES.length];
    case 2: return TWO_BED_IMAGES[index % TWO_BED_IMAGES.length];
    case 3: return THREE_BED_IMAGES[index % THREE_BED_IMAGES.length];
    case 4: return FOUR_BED_IMAGES[index % FOUR_BED_IMAGES.length];
    default: return THREE_BED_IMAGES[0];
  }
}

// ---------------------------------------------------------------------------
// Zillow response → PropertyCard mapper
// ---------------------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function mapZillowToPropertyCard(property: any, index: number): PropertyCard {
  const zillowPhoto = property.media?.propertyPhotoLinks?.highResolutionLink;

  return {
    zpid: property.zpid,

    // Address
    street_address: property.address?.streetAddress,
    city: property.address?.city,
    state: property.address?.state,
    zip_code: property.address?.zipcode,

    // Coordinates
    latitude: property.location?.latitude,
    longitude: property.location?.longitude,

    // Listing info
    price: property.price?.value,
    price_per_sqft: property.price?.pricePerSquareFoot,
    price_change: property.price?.priceChange,
    price_changed_date: property.price?.changedDate
      ? new Date(property.price.changedDate).toISOString()
      : undefined,
    listing_status: property.listing?.listingStatus,
    days_on_zillow: property.daysOnZillow,
    listing_date: property.listingDateTimeOnZillow
      ? new Date(property.listingDateTimeOnZillow).toISOString()
      : undefined,

    // Property details
    property_type: property.propertyType,
    beds: property.bedrooms,
    baths: property.bathrooms,
    sqft: property.livingArea,
    lot_size: property.lotSizeWithUnit?.lotSize,
    lot_size_unit: property.lotSizeWithUnit?.lotSizeUnit,
    year_built: property.yearBuilt,
    is_new_construction: property.listing?.listingSubType?.isNewConstruction,

    // Estimates
    zestimate: property.estimates?.zestimate,
    rent_zestimate: property.estimates?.rentZestimate,

    // Tax
    tax_assessed_value: property.taxAssessment?.taxAssessedValue,
    tax_assessment_year: property.taxAssessment?.taxAssessmentYear,

    // Media flags
    has_vr_model: property.media?.hasVRModel,
    has_videos: property.media?.hasVideos,
    has_floor_plan: property.hasFloorPlan,
    is_showcase_listing: property.isShowcaseListing,

    // Open house
    open_house_start: property.openHouseShowingList?.[0]?.startTime
      ? new Date(property.openHouseShowingList[0].startTime).toISOString()
      : undefined,
    open_house_end: property.openHouseShowingList?.[0]?.endTime
      ? new Date(property.openHouseShowingList[0].endTime).toISOString()
      : undefined,

    // Broker
    broker_name: property.propertyDisplayRules?.mls?.brokerName,

    // Use Zillow photo if available, otherwise fall back to local assets
    image: zillowPhoto ?? getFallbackImage(property.bedrooms ?? 3, index),
  };
}

// ---------------------------------------------------------------------------
// Fetch live properties from your backend (which calls the Zillow API)
// ---------------------------------------------------------------------------
export async function fetchProperties(): Promise<PropertyCard[]> {
  const response = await fetch(`${API_BASE}/zillow/search`);

  if (!response.ok) {
    throw new Error(`Failed to fetch properties: ${response.statusText}`);
  }

  const data = await response.json();

  // Zillow wraps results under a `results` array — adjust key if your
  // backend reshapes the response differently
  const results = data.results ?? data;

  return results.map(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (property: any, index: number) => mapZillowToPropertyCard(property, index)
  );
}

File: backend/app/routers/properties.py
from typing import Annotated, List
from sqlalchemy.orm import Session
from fastapi import APIRouter, Depends, HTTPException
from app.schemas.properties import PropertiesCreate, PropertiesOut
from app.db.models.properties import Properties
from app.deps.db import get_db

router = APIRouter(tags=["properties"])


@router.get("/properties/{user_id}", response_model=List[PropertiesOut])
def get_properties(
    user_id: str,
    db: Annotated[Session, Depends(get_db)],
):
    return db.query(Properties).filter(Properties.user_id == user_id).all()


@router.post("/properties", response_model=PropertiesOut)
def create_properties(
    req: PropertiesCreate,
    db: Annotated[Session, Depends(get_db)],
):
    new_property = Properties(
        user_id=req.user_id,
        liked=req.liked,
        zpid=req.zpid,

        # Address
        street_address=req.street_address,
        city=req.city,
        state=req.state,
        zip_code=req.zip_code,

        # Coordinates
        latitude=req.latitude,
        longitude=req.longitude,

        # Listing info
        price=req.price,
        price_per_sqft=req.price_per_sqft,
        price_change=req.price_change,
        price_changed_date=req.price_changed_date,
        listing_status=req.listing_status,
        days_on_zillow=req.days_on_zillow,
        listing_date=req.listing_date,

        # Property details
        property_type=req.property_type,
        beds=req.beds,
        baths=req.baths,
        sqft=req.sqft,
        lot_size=req.lot_size,
        lot_size_unit=req.lot_size_unit,
        year_built=req.year_built,
        is_new_construction=req.is_new_construction,

        # Estimates
        zestimate=req.zestimate,
        rent_zestimate=req.rent_zestimate,

        # Tax
        tax_assessed_value=req.tax_assessed_value,
        tax_assessment_year=req.tax_assessment_year,

        # Media flags
        has_vr_model=req.has_vr_model,
        has_videos=req.has_videos,
        has_floor_plan=req.has_floor_plan,
        is_showcase_listing=req.is_showcase_listing,

        # Open house
        open_house_start=req.open_house_start,
        open_house_end=req.open_house_end,

        # Broker
        broker_name=req.broker_name,

        # Thumbnail
        photo_url=req.photo_url,
    )
    db.add(new_property)
    db.commit()
    db.refresh(new_property)
    return new_property


@router.delete("/properties/{property_id}")
def delete_property(
    property_id: int,
    db: Annotated[Session, Depends(get_db)],
):
    prop = db.query(Properties).filter(Properties.id == property_id).first()
    if not prop:
        raise HTTPException(status_code=404, detail="Property not found")
    db.delete(prop)
    db.commit()
    return {"ok": True}
